pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/GrpcBrokerRequestHandler.java

```diff
@@ -0,0 +1,198 @@
+      responseMap.putAll(serverResponses);
+    }
+    BrokerResponseNative brokerResponse = _streamingReduceService.reduceOnStreamResponse(
+        originalBrokerRequest, responseMap, timeoutMs, _brokerMetrics);
+    return brokerResponse;
+  }
+
+  /**
+   * Query pinot server for data table.
+   */
+  public Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> streamingQueryToPinotServer(
+      final long requestId,
+      final String brokerHost,
+      final String rawTableName,
+      final TableType tableType,
+      BrokerRequest brokerRequest,
+      Map<ServerInstance, List<String>> routingTable,
+      long connectionTimeoutInMillis,
+      boolean ignoreEmptyResponses,
+      int pinotRetryCount) {
+    // Retries will all hit the same server because the routing decision has already been made by the pinot broker
+    Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> serverResponseMap = new HashMap<>();
+    for (Map.Entry<ServerInstance, List<String>> routingEntry : routingTable.entrySet()) {
+      ServerInstance serverInstance = routingEntry.getKey();
+      List<String> segments = routingEntry.getValue();
+      String serverHost = serverInstance.getHostname();
+      int port = serverInstance.getGrpcPort();
+      // ensure concurrent request count from the same grpc server cannot exceed maximum
+      if (!_concurrentQueriesCountMap.containsKey(serverHost)) {
+        _concurrentQueriesCountMap.put(serverHost, new AtomicInteger(0));
+      }
+      int concurrentQueryNum = _concurrentQueriesCountMap.get(serverHost).get();
+      if (concurrentQueryNum > _maxBacklogPerServer) {
+        ProcessingException processingException = new ProcessingException(QueryException.UNKNOWN_ERROR_CODE);
+        processingException.setMessage("Reaching server query max backlog size is - " + _maxBacklogPerServer);
+        throw new RuntimeException(processingException);
+      }
+      // TODO: enable throttling on per host bases.
+//      _concurrentQueriesCountMap.get(serverHost).incrementAndGet();
+      Iterator<Server.ServerResponse> streamingResponse = _streamingQueryClient.submit(serverHost, port,
+          new GrpcRequestBuilder()
+              .setSegments(segments)
+              .setBrokerRequest(brokerRequest)
+              .setEnableStreaming(true));
+      serverResponseMap.put(serverInstance.toServerRoutingInstance(tableType), streamingResponse);
+    }
+    return serverResponseMap;
+  }
+
+  private static long makeGrpcRequestId(long requestId, int requestAttemptId) {
+    return requestId * DEFAULT_MAXIMUM_REQUEST_ATTEMPT + requestAttemptId;
+  }
+
+  // return empty config for now
+  private GrpcQueryClient.Config buildGrpcQueryClientConfig(PinotConfiguration config) {
+    return new GrpcQueryClient.Config();
+  }
+
+  public static class PinotStreamingQueryClient {
+    private final Map<String, GrpcQueryClient> _grpcQueryClientMap = new HashMap<>();
```

Comment:
This is not thread safe. We need to use a concurrent map, and call `computeIfAbsent()`